src/selfplay/tournament.cc

```diff
@@ -205,103 +189,156 @@ void SelfPlayTournament::PlayOneGame(int game_number) {
-          }
+        [this, game_number, pl_idx, player1_black,
+         verbose_thinking](const std::vector<ThinkingInfo>& infos) {
           if (verbose_thinking) {
+            std::vector<ThinkingInfo> rich_info = infos;
+            for (auto& info : rich_info) {
+              info.player = pl_idx + 1;
+              info.is_black = player1_black ? pl_idx == 0 : pl_idx != 0;
+              info.game_id = game_number;
+            }
             info_callback_(rich_info);
-          } else {
-            // In non-verbose mode, remember the last "info" messages.
-            last_thinking_info = std::move(rich_info);
           }
         };
   }
 
-  // Iterator to store the game in. Have to keep it so that later we can
-  // delete it. Need to expose it in games_ member variable only because
-  // of possible Abort() that should stop them all.
-  std::list<std::unique_ptr<SelfPlayGame>>::iterator game_iter;
-  {
-    Mutex::Lock lock(mutex_);
-    games_.emplace_front(
-        std::make_unique<SelfPlayGame>(options[0], options[1], kShareTree));
-    game_iter = games_.begin();
-  }
-  auto& game = **game_iter;
-
   // If kResignPlaythrough == 0, then this comparison is unconditionally true
   bool enable_resign = Random::Get().GetFloat(100.0f) >= kResignPlaythrough;
 
-  // PLAY GAME!
-  game.Play(kThreads[color_idx[0]], kThreads[color_idx[1]], kTraining,
-            enable_resign);
-
-  // If game was aborted, it's still undecided.
-  if (game.GetGameResult() != GameResult::UNDECIDED) {
-    // Game callback.
-    GameInfo game_info;
-    game_info.game_result = game.GetGameResult();
-    game_info.is_black = player1_black;
-    game_info.game_id = game_number;
-    game_info.moves = game.GetMoves();
-    if (!enable_resign) {
-      game_info.min_false_positive_threshold =
-          game.GetWorstEvalForWinnerOrDraw();
-    }
-    if (kTraining) {
-      TrainingDataWriter writer(game_number);
-      game.WriteTrainingData(&writer);
-      writer.Finalize();
-      game_info.training_filename = writer.GetFileName();
-    }
-    game_callback_(game_info);
+  return std::make_unique<SelfPlayGame>(game_number, options[color_idx[0]],
+                                        options[color_idx[1]], kShareTree,
+                                        enable_resign);
+}
 
-    // Update tournament stats.
-    {
-      Mutex::Lock lock(mutex_);
-      int result = game.GetGameResult() == GameResult::DRAW
-                       ? 1
-                       : game.GetGameResult() == GameResult::WHITE_WON ? 0 : 2;
-      if (player1_black) result = 2 - result;
-      ++tournament_info_.results[result][player1_black ? 1 : 0];
-      tournament_callback_(tournament_info_);
-    }
+// Called when the SelfPlayGame is finished, so that it's processed.
+void SelfPlayTournament::SendGameReport(const SelfPlayGame& game) {
+  const int game_number = game.GetGameNumber();
+  const bool player1_black = (game_number % 2) != first_game_is_flipped_;
```

Comment:
Maybe it would be cleaner if SelfPlayGame knew which was which and carried that info around rather than recomputing it in multiple places?

Reply:
Done.